CIP3/CIP3.md

```diff
@@ -0,0 +1,234 @@
+
+More specifically, the generation is a function from an initial seed to an extended private key (abbrev. XPrv) composed of:
+
+- 64 bytes: an extended Ed25519 secret key composed of:
+  - 32 bytes: Ed25519 curve scalar from which few bits have been tweaked according to ED25519-BIP32
+  - 32 bytes: Ed25519 binary blob used as IV for signing
+- 32 bytes: chain code for allowing secure child key derivation
+
+#### History
+
+Throughout the years, Cardano has used different styles of master key generation:
+
+|  Name          |  Created by |  Address prefix in Byron |  Is deprecated? |
+|----------------|-------------|--------------------------|-----------------|
+|  Byron         |  Daedalus   |  Ddz                     |  Yes            |
+|  Icarus        |  Yoroi      |  Ae2                     |  No             |
+|  Icarus-Trezor |  Trezor     |  Ae2                     |  No             |
+|  Ledger        |  Ledger     |  Ae2                     |  No             |
+
+# Rationale
+
+This CIP is merely to document the existing standards and not to provide rationales for the various methods used.
+
+However, you can learn more at the following links:
+
+- [Adrestia documentation](https://input-output-hk.github.io/adrestia/docs/key-concepts/hierarchical-deterministic-wallets/)
+- [SLIP-0010](https://github.com/satoshilabs/slips/blob/master/slip-0010.md)
+- [SLIP-0023](https://github.com/satoshilabs/slips/blob/master/slip-0023.md)
+
+## Reference implementations
+
+We provide a description, reference implementation and test vectors for each algorithm
+
+#### Byron
+
+```js
+function generateMasterKey(seed) {
+    return hashRepeatedly(seed, 1);
+}
+
+function hashRepeatedly(key, i) {
+    (iL, iR) = HMAC
+        ( hash=SHA512
+        , key=key
+        , message="Root Seed Chain " + UTF8NFKD(i)
+        );
+
+    let prv = tweakBits(SHA512(iL));
+
+    if (prv[31] & 0b0010_0000) {
+        return hashRepeatedly(key, i+1);
+    }
+
+    return (prv + iR);
+}
+
+function tweakBits(data) {
+    // * clear the lowest 3 bits
+    // * clear the highest bit
+    // * set the highest 2nd bit
+    data[0]  &= 0b1111_1000;
+    data[31] &= 0b0111_1111;
+    data[31] |= 0b0100_0000;
+
+    return data;
+}
+```
+
+##### Test vectors
+
+```
+mnemonic: install neither high supreme hurdle tissue excite lava census harbor purpose shine
+master key: 309c18090ec559d4c8e377b9b19b84804a4fc430e862ef99d97f1cceccad124d92d7336d2d598affefc05c806fe2ff71cfc0fb78bfdb539178dbfcc567a04e358077d66a909e93f4577b5784b8c1c4ff3c77e706449f1c21793a79fc2abb68a9
+```
+
+#### Icarus
+
+Icarus master key generation style supports setting an extra password as an arbitrary byte array of any size. This password acts as a second factor applied to cryptographic key retrieval. When the seed comes from an encoded recovery phrase, the password can therefore be used to add extra protection in case where the recovery phrase were to be exposed.
+
+```js
```

Comment:
I inline all the reference implementation with test vectors directly in the specification. However, possibly these should be uploaded as Gist or separate JS libraries since probably this CIP is better as an outline of key generation types instead of growing in size over time if new generation schemes get added. This would be more readable and easier to update.

Reply:
I think there's no wrong in principle in adding these files directly in the same folder and version them alongside the main CIP document?